File: go/bandwidth/monitor.go
// SPDX-License-Identifier: MIT

package bandwidth

import (
	"math"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

// NodeActivityMonitor is the UI-facing sampler of the ABMF.
//
// It runs a background goroutine at SampleIntervalMs (default 500 ms).
// Each tick computes ingress/egress rates from atomic byte counters, reads
// per-transport estimates from registered BandwidthEstimators, and publishes
// a NodeActivitySnapshot.
//
// Rate computation: byte deltas are divided by the elapsed wall-clock interval.
// The sample interval acts as the averaging window (allocation-free on the hot path).
//
// All methods are safe for concurrent use.
type NodeActivityMonitor struct {
	// SampleIntervalMs controls how often the monitor re-samples (milliseconds).
	// Default: 500. Must be set before Start() for best results; changing after
	// Start() takes effect on the next tick.
	SampleIntervalMs int

	// IdleThresholdSeconds controls how long without traffic before a transport
	// is considered idle. Default: 5.
	IdleThresholdSeconds int

	mu         sync.RWMutex
	transports map[string]*transportEntry // keyed by lower-case name

	// lastSeenPeerMs maps peerUhid → last-seen Unix ms. A peer is "active" if it
	// had ingress or egress within IdleThresholdSeconds. Populated only by the
	// peer-aware RecordIngressFromPeer/RecordEgressToPeer methods; the
	// transport-only RecordIngress/RecordEgress do not contribute (the caller did
	// not supply a peer). Stale entries are pruned each tick so the map stays
	// bounded by the count of recently-active peers, not the lifetime peer set.
	peerMu         sync.Mutex
	lastSeenPeerMs map[string]int64

	// current is the last published snapshot (pointer; swapped atomically).
	currentMu sync.RWMutex
	current   *NodeActivitySnapshot

	// subscription management
	subMu   sync.Mutex
	subs    map[uint64]func(NodeActivitySnapshot)
	nextSub uint64

	stopCh chan struct{}
	doneCh chan struct{}
}

type transportEntry struct {
	name      string
	estimator *BandwidthEstimator

	// Atomic byte counters reset every tick.
	ingressBytes int64
	egressBytes  int64

	// Last egress timestamp in Unix ms.
	lastEgressMs int64
}

// NewNodeActivityMonitor returns a monitor ready to be configured and started.
func NewNodeActivityMonitor() *NodeActivityMonitor {
	return &NodeActivityMonitor{
		SampleIntervalMs:     500,
		IdleThresholdSeconds: 5,
		transports:           make(map[string]*transportEntry),
		lastSeenPeerMs:       make(map[string]int64),
		current:              offlineSnapshot(),
		subs:                 make(map[uint64]func(NodeActivitySnapshot)),
	}
}

// Register adds a transport's estimator so its activity is included in snapshots.
func (m *NodeActivityMonitor) Register(name string, estimator *BandwidthEstimator) {
	key := strings.ToLower(name)
	entry := &transportEntry{
		name:         name,
		estimator:    estimator,
		lastEgressMs: time.Now().UnixMilli(),
	}
	m.mu.Lock()
	m.transports[key] = entry
	m.mu.Unlock()
}

// RecordIngress records inbound bytes on the named transport.
// Safe to call from any goroutine; uses atomic addition.
func (m *NodeActivityMonitor) RecordIngress(transport string, bytes int) {
	key := strings.ToLower(transport)
	m.mu.RLock()
	entry, ok := m.transports[key]
	m.mu.RUnlock()
	if ok {
		atomic.AddInt64(&entry.ingressBytes, int64(bytes))
	}
}

// RecordEgress records outbound bytes on the named transport.
// Safe to call from any goroutine; uses atomic addition.
func (m *NodeActivityMonitor) RecordEgress(transport string, bytes int) {
	key := strings.ToLower(transport)
	m.mu.RLock()
	entry, ok := m.transports[key]
	m.mu.RUnlock()
	if ok {
		atomic.AddInt64(&entry.egressBytes, int64(bytes))
		atomic.StoreInt64(&entry.lastEgressMs, time.Now().UnixMilli())
	}
}

// RecordIngressFromPeer records inbound bytes on the named transport from a
// specific peer. It updates the transport counters via RecordIngress and tracks
// the peer for the NodeActivitySnapshot.ActivePeers count.
// Safe to call from any goroutine.
func (m *NodeActivityMonitor) RecordIngressFromPeer(transport, peerUhid string, bytes int) {
	m.RecordIngress(transport, bytes)
	m.recordPeerSeen(peerUhid)
}

// RecordEgressToPeer records outbound bytes on the named transport to a specific
// peer. It updates the transport counters via RecordEgress and tracks the peer
// for the NodeActivitySnapshot.ActivePeers count.
// Safe to call from any goroutine.
func (m *NodeActivityMonitor) RecordEgressToPeer(transport, peerUhid string, bytes int) {
	m.RecordEgress(transport, bytes)
	m.recordPeerSeen(peerUhid)
}

// recordPeerSeen stamps peerUhid with the current Unix ms. Empty UHIDs are ignored.
func (m *NodeActivityMonitor) recordPeerSeen(peerUhid string) {
	if peerUhid == "" {
		return
	}
	now := time.Now().UnixMilli()
	m.peerMu.Lock()
	m.lastSeenPeerMs[peerUhid] = now
	m.peerMu.Unlock()
}

// Start launches the background sampling goroutine.
// Calling Start more than once has no effect.
func (m *NodeActivityMonitor) Start() {
	m.mu.Lock()
	if m.stopCh != nil {
		m.mu.Unlock()
		return
	}
	m.stopCh = make(chan struct{})
	m.doneCh = make(chan struct{})
	m.mu.Unlock()

	go m.loop()
}

// Stop halts the background goroutine and waits for it to exit.
func (m *NodeActivityMonitor) Stop() {
	m.mu.Lock()
	stopCh := m.stopCh
	doneCh := m.doneCh
	m.stopCh = nil
	m.mu.Unlock()

	if stopCh != nil {
		close(stopCh)
		<-doneCh
	}
}

// Current returns the most recent snapshot.
// Never nil after construction — initialised to an Offline snapshot with zero rates.
func (m *NodeActivityMonitor) Current() NodeActivitySnapshot {
	m.currentMu.RLock()
	s := *m.current
	m.currentMu.RUnlock()
	return s
}

// Subscribe registers a callback that fires every time a new snapshot is published.
// The returned function removes the subscription; calling it more than once is safe.
func (m *NodeActivityMonitor) Subscribe(fn func(NodeActivitySnapshot)) func() {
	m.subMu.Lock()
	id := m.nextSub
	m.nextSub++
	m.subs[id] = fn
	m.subMu.Unlock()

	return func() {
		m.subMu.Lock()
		delete(m.subs, id)
		m.subMu.Unlock()
	}
}

// loop is the background sampling goroutine.
func (m *NodeActivityMonitor) loop() {
	// Capture channels once at start; they are set before Start() returns.
	m.mu.Lock()
	stopCh := m.stopCh
	doneCh := m.doneCh
	m.mu.Unlock()

	defer close(doneCh)

	lastTickMs := time.Now().UnixMilli()
	for {
		interval := time.Duration(m.SampleIntervalMs) * time.Millisecond

		select {
		case <-stopCh:
			return
		case <-time.After(interval):
		}

		nowMillis := time.Now().UnixMilli()
		elapsedSec := math.Max(0.001, float64(nowMillis-lastTickMs)/1000.0)
		lastTickMs = nowMillis

		m.tick(elapsedSec, nowMillis)
	}
}

// tick computes one snapshot and publishes it.
func (m *NodeActivityMonitor) tick(elapsedSec float64, nowMillis int64) {
	m.mu.RLock()
	idleThreshMs := int64(m.IdleThresholdSeconds) * 1000
	entries := make([]*transportEntry, 0, len(m.transports))
	for _, e := range m.transports {
		entries = append(entries, e)
	}
	m.mu.RUnlock()

	// Count distinct peers active within the idle window; prune stale entries so
	// the map stays bounded by recently-active peers rather than the lifetime set.
	activePeers := 0
	m.peerMu.Lock()
	for uhid, lastSeen := range m.lastSeenPeerMs {
		if nowMillis-lastSeen < idleThreshMs {
			activePeers++
		} else {
			delete(m.lastSeenPeerMs, uhid)
		}
	}
	m.peerMu.Unlock()

	var totalIngress, totalEgress int64
	var activeTransports int
	var transportSnaps []TransportActivitySnapshot

	for _, entry := range entries {
		ingressDelta := atomic.SwapInt64(&entry.ingressBytes, 0)
		egressDelta := atomic.SwapInt64(&entry.egressBytes, 0)

		ingressBps := int64(float64(ingressDelta) * 8.0 / elapsedSec)
		egressBps := int64(float64(egressDelta) * 8.0 / elapsedSec)

		s := entry.estimator.CurrentSample()
		utilFraction := 0.0
		if s.BtlBwBps > 0 {
			utilFraction = clampFloat64(float64(egressBps)/float64(s.BtlBwBps), 0.0, 1.0)
		}

		lastEgress := atomic.LoadInt64(&entry.lastEgressMs)
		isRecent := (nowMillis - lastEgress) < idleThreshMs
		state := computeTransportState(egressBps, ingressBps, s, isRecent)

		if state != NodeOffline && state != NodeIdle {
			activeTransports++
		}

		totalIngress += ingressBps
		totalEgress += egressBps

		transportSnaps = append(transportSnaps, TransportActivitySnapshot{
			TransportName:       entry.name,
			IsAvailable:         true,
			IngressBps:          ingressBps,
			EgressBps:           egressBps,
			Srtt:                s.Srtt,
			BtlBwBps:            s.BtlBwBps,
			UtilizationFraction: utilFraction,
			State:               state,
			Confidence:          s.Confidence,
		})
	}

	nodeState := computeNodeState(transportSnaps)

	primaryName := ""
	var primaryBps int64 = -1
	for _, ts := range transportSnaps {
		if ts.EgressBps > primaryBps {
			primaryBps = ts.EgressBps
			primaryName = ts.TransportName
		}
	}
	if nodeState == NodeOffline || nodeState == NodeIdle {
		primaryName = ""
	}

	snapshot := &NodeActivitySnapshot{
		State:                nodeState,
		IngressBps:           totalIngress,
		EgressBps:            totalEgress,
		ActivePeers:          activePeers,
		ActiveTransports:     activeTransports,
		Transports:           transportSnaps,
		PrimaryTransportName: primaryName,
		Timestamp:            time.Now().UTC(),
	}

	prev := m.Current()
	m.currentMu.Lock()
	m.current = snapshot
	m.currentMu.Unlock()

	// Fire subscribers.
	m.subMu.Lock()
	cbs := make([]func(NodeActivitySnapshot), 0, len(m.subs))
	for _, fn := range m.subs {
		cbs = append(cbs, fn)
	}
	m.subMu.Unlock()

	s := *snapshot
	for _, fn := range cbs {
		safeCall(fn, s)
	}

	// Suppress if identical (state/rates/transport count unchanged).
	_ = prev // available for change detection if callers need it
}

// ── State computation ────────────────────────────────────────────────────────

func computeTransportState(
	egressBps, ingressBps int64,
	s BandwidthSample,
	isRecent bool,
) NodeActivityState {
	if !isRecent && egressBps == 0 && ingressBps == 0 {
		return NodeIdle
	}
	if egressBps == 0 && ingressBps == 0 {
		return NodeIdle
	}
	if s.LossRate > 0.05 {
		return NodeDegraded
	}
	util := 0.0
	if s.BtlBwBps > 0 {
		util = float64(egressBps) / float64(s.BtlBwBps)
	}
	if util >= 0.5 {
		return NodeBusy
	}
	return NodeActive
}

func computeNodeState(transports []TransportActivitySnapshot) NodeActivityState {
	if len(transports) == 0 {
		return NodeOffline
	}
	hasDegraded, hasBusy, hasActive, allOffline := false, false, false, true
	for _, t := range transports {
		switch t.State {
		case NodeDegraded:
			hasDegraded = true
			allOffline = false
		case NodeBusy:
			hasBusy = true
			allOffline = false
		case NodeActive:
			hasActive = true
			allOffline = false
		case NodeIdle:
			allOffline = false
		}
	}
	switch {
	case hasDegraded:
		return NodeDegraded
	case hasBusy:
		return NodeBusy
	case hasActive:
		return NodeActive
	case allOffline:
		return NodeOffline
	default:
		return NodeIdle
	}
}

// offlineSnapshot is the zero-value snapshot returned before any data flows.
func offlineSnapshot() *NodeActivitySnapshot {
	return &NodeActivitySnapshot{
		State:     NodeOffline,
		Timestamp: time.Now().UTC(),
	}
}

// safeCall invokes fn, swallowing any panics so subscriber errors cannot kill the loop.
func safeCall(fn func(NodeActivitySnapshot), s NodeActivitySnapshot) {
	defer func() { recover() }() //nolint:errcheck
	fn(s)
}

File: rust/src/bandwidth/monitor.rs
// SPDX-License-Identifier: MIT

//! Node activity monitor — UI-facing layer of the ABMF.
//!
//! Ports `src/AetherNet.Transport/Bandwidth/NodeActivityMonitor.cs`.

use std::collections::HashMap;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use super::estimator::BandwidthEstimator;
use super::models::{NodeActivitySnapshot, NodeActivityState, TransportActivitySnapshot};

// ── Traffic accumulator ───────────────────────────────────────────────────────

struct TransportTraffic {
    ingress_bytes: AtomicI64,
    egress_bytes: AtomicI64,
    last_egress_ms: AtomicI64,
}

impl TransportTraffic {
    fn new() -> Self {
        let now_ms = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or(Duration::ZERO)
            .as_millis() as i64;
        Self {
            ingress_bytes: AtomicI64::new(0),
            egress_bytes: AtomicI64::new(0),
            last_egress_ms: AtomicI64::new(now_ms),
        }
    }
}

// ── Shared state ──────────────────────────────────────────────────────────────

struct Shared {
    transports: HashMap<String, (Arc<Mutex<BandwidthEstimator>>, Arc<TransportTraffic>)>,
    // Maps peer_uhid → last-seen Unix ms. A peer is "active" if it had ingress or
    // egress within idle_threshold_ms. Populated only by the peer-aware
    // record_ingress_from_peer / record_egress_to_peer methods; the transport-only
    // methods do not contribute (the caller did not supply a peer). Stale entries
    // are pruned each tick so the map stays bounded by the count of recently-active
    // peers, not the lifetime peer set.
    last_seen_peer_ms: HashMap<String, i64>,
    sample_interval_ms: u64,
    idle_threshold_ms: i64,
    current: NodeActivitySnapshot,
    subscribers: Vec<Box<dyn Fn(NodeActivitySnapshot) + Send + 'static>>,
    running: bool,
}

// ── Monitor ───────────────────────────────────────────────────────────────────

/// Runs a background thread sampling ingress/egress rates and publishing
/// `NodeActivitySnapshot` objects at a configurable interval (default 500 ms).
pub struct NodeActivityMonitor {
    shared: Arc<Mutex<Shared>>,
}

impl NodeActivityMonitor {
    pub fn new() -> Self {
        let initial = offline_snapshot();
        Self {
            shared: Arc::new(Mutex::new(Shared {
                transports: HashMap::new(),
                last_seen_peer_ms: HashMap::new(),
                sample_interval_ms: 500,
                idle_threshold_ms: 5_000,
                current: initial,
                subscribers: Vec::new(),
                running: false,
            })),
        }
    }

    // ── Configuration ─────────────────────────────────────────────────────────

    pub fn set_sample_interval_ms(&self, ms: u64) {
        self.shared.lock().unwrap().sample_interval_ms = ms.clamp(100, 60_000);
    }

    pub fn set_idle_threshold_seconds(&self, secs: i64) {
        self.shared.lock().unwrap().idle_threshold_ms = secs.clamp(1, 300) * 1000;
    }

    // ── Registration ──────────────────────────────────────────────────────────

    /// Register a transport estimator so its activity is included in snapshots.
    pub fn register(&self, name: &str, estimator: Arc<Mutex<BandwidthEstimator>>) {
        let traffic = Arc::new(TransportTraffic::new());
        self.shared
            .lock()
            .unwrap()
            .transports
            .insert(name.to_string(), (estimator, traffic));
    }

    /// Record inbound bytes on a transport.
    pub fn record_ingress(&self, transport: &str, bytes: i32) {
        let g = self.shared.lock().unwrap();
        if let Some((_, traffic)) = g.transports.get(transport) {
            traffic.ingress_bytes.fetch_add(bytes as i64, Ordering::Relaxed);
        }
    }

    /// Record outbound bytes on a transport.
    pub fn record_egress(&self, transport: &str, bytes: i32) {
        let now_ms = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or(Duration::ZERO)
            .as_millis() as i64;
        let g = self.shared.lock().unwrap();
        if let Some((_, traffic)) = g.transports.get(transport) {
            traffic.egress_bytes.fetch_add(bytes as i64, Ordering::Relaxed);
            traffic.last_egress_ms.store(now_ms, Ordering::Relaxed);
        }
    }

    /// Record inbound bytes on a transport from a specific peer.
    /// Tracks the peer for the `NodeActivitySnapshot::active_peers` count.
    pub fn record_ingress_from_peer(&self, transport: &str, peer_uhid: &str, bytes: i32) {
        self.record_ingress(transport, bytes);
        if !peer_uhid.is_empty() {
            let now_ms = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or(Duration::ZERO)
                .as_millis() as i64;
            self.shared
                .lock()
                .unwrap()
                .last_seen_peer_ms
                .insert(peer_uhid.to_string(), now_ms);
        }
    }

    /// Record outbound bytes on a transport to a specific peer.
    /// Tracks the peer for the `NodeActivitySnapshot::active_peers` count.
    pub fn record_egress_to_peer(&self, transport: &str, peer_uhid: &str, bytes: i32) {
        self.record_egress(transport, bytes);
        if !peer_uhid.is_empty() {
            let now_ms = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or(Duration::ZERO)
                .as_millis() as i64;
            self.shared
                .lock()
                .unwrap()
                .last_seen_peer_ms
                .insert(peer_uhid.to_string(), now_ms);
        }
    }

    // ── Subscription ─────────────────────────────────────────────────────────

    /// Subscribe a callback fired on every snapshot (heartbeat).
    pub fn subscribe(&self, cb: Box<dyn Fn(NodeActivitySnapshot) + Send + 'static>) {
        self.shared.lock().unwrap().subscribers.push(cb);
    }

    // ── Lifecycle ────────────────────────────────────────────────────────────

    /// Start the background sampling loop.
    pub fn start(&self) {
        {
            let mut g = self.shared.lock().unwrap();
            if g.running {
                return;
            }
            g.running = true;
        }
        let shared = Arc::clone(&self.shared);
        thread::Builder::new()
            .name("abmf-monitor".to_string())
            .spawn(move || run_loop(shared))
            .expect("failed to spawn monitor thread");
    }

    /// Stop the background sampling loop.
    pub fn stop(&self) {
        self.shared.lock().unwrap().running = false;
    }

    // ── Query ─────────────────────────────────────────────────────────────────

    /// The most recent snapshot. Thread-safe.
    pub fn current(&self) -> NodeActivitySnapshot {
        self.shared.lock().unwrap().current.clone()
    }
}

impl Default for NodeActivityMonitor {
    fn default() -> Self {
        Self::new()
    }
}

// ── Background loop ───────────────────────────────────────────────────────────

fn run_loop(shared: Arc<Mutex<Shared>>) {
    let mut last_tick_ms = now_ms();

    loop {
        let interval_ms = {
            let g = shared.lock().unwrap();
            if !g.running {
                break;
            }
            g.sample_interval_ms
        };

        thread::sleep(Duration::from_millis(interval_ms));

        let now = now_ms();
        let elapsed_sec = ((now - last_tick_ms) / 1000.0).max(0.001);
        last_tick_ms = now;

        let snapshot = {
            let mut g = shared.lock().unwrap();
            if !g.running {
                break;
            }
            build_snapshot(&mut g, elapsed_sec, now)
        };

        let (prev_state, prev_total, prev_active_transports) = {
            let g = shared.lock().unwrap();
            (
                g.current.state,
                g.current.total_bps(),
                g.current.active_transports,
            )
        };

        // Publish snapshot: update current, then call subscribers WITHOUT holding the lock.
        // We take a snapshot of subscriber count, store the current snapshot, release the lock,
        // then dispatch callbacks individually (re-acquiring the lock per call to safely index in).
        let sub_count = {
            let mut g = shared.lock().unwrap();
            g.current = snapshot.clone();
            g.subscribers.len()
        };

        for idx in 0..sub_count {
            let cb_snapshot = snapshot.clone();
            // Re-acquire briefly to call the callback.  Callbacks must not call back into
            // the monitor or they will deadlock on this same mutex.
            let g = shared.lock().unwrap();
            if let Some(cb) = g.subscribers.get(idx) {
                cb(cb_snapshot);
            }
        }

        // Detect change for potential future SnapshotChanged event.
        let _changed = snapshot.state != prev_state
            || (snapshot.total_bps() - prev_total).abs() > 1_000
            || snapshot.active_transports != prev_active_transports;
    }
}

fn build_snapshot(g: &mut Shared, elapsed_sec: f64, now_ms: f64) -> NodeActivitySnapshot {
    let mut transport_snapshots: Vec<TransportActivitySnapshot> =
        Vec::with_capacity(g.transports.len());
    let mut total_ingress: i64 = 0;
    let mut total_egress: i64 = 0;
    let mut active_transports: i32 = 0;

    // Count distinct peers active within the idle window; prune stale entries
    // so the map stays bounded by recently-active peers.
    let now_i64 = now_ms as i64;
    let idle_threshold_ms = g.idle_threshold_ms;
    g.last_seen_peer_ms
        .retain(|_, last_seen| now_i64 - *last_seen < idle_threshold_ms);
    let active_peers = g.last_seen_peer_ms.len() as i32;

    for (name, (estimator, traffic)) in &g.transports {
        let ingress_delta = traffic.ingress_bytes.swap(0, Ordering::Relaxed);
        let egress_delta = traffic.egress_bytes.swap(0, Ordering::Relaxed);
        let ingress_bps = (ingress_delta as f64 * 8.0 / elapsed_sec) as i64;
        let egress_bps = (egress_delta as f64 * 8.0 / elapsed_sec) as i64;

        let sample = estimator.lock().unwrap().current_sample();
        let util_fraction = if sample.btl_bw_bps > 0 {
            (egress_bps as f64 / sample.btl_bw_bps as f64).clamp(0.0, 1.0)
        } else {
            0.0
        };

        let last_egress = traffic.last_egress_ms.load(Ordering::Relaxed);
        let is_recent = (now_ms as i64 - last_egress) < g.idle_threshold_ms;
        let state = compute_transport_state(egress_bps, ingress_bps, &sample, is_recent);

        if !matches!(state, NodeActivityState::Offline | NodeActivityState::Idle) {
            active_transports += 1;
        }

        total_ingress += ingress_bps;
        total_egress += egress_bps;

        transport_snapshots.push(TransportActivitySnapshot {
            transport_name: name.clone(),
            is_available: true,
            ingress_bps,
            egress_bps,
            srtt: sample.srtt,
            btl_bw_bps: sample.btl_bw_bps,
            utilization_fraction: util_fraction,
            state,
            confidence: sample.confidence,
        });
    }

    let node_state = compute_node_state(&transport_snapshots);
    let primary = transport_snapshots
        .iter()
        .max_by_key(|t| t.egress_bps)
        .filter(|t| t.egress_bps > 0)
        .map(|t| t.transport_name.clone());

    NodeActivitySnapshot {
        state: node_state,
        ingress_bps: total_ingress,
        egress_bps: total_egress,
        active_peers,
        active_transports,
        transports: transport_snapshots,
        primary_transport_name: primary,
        timestamp: SystemTime::now(),
    }
}

fn compute_transport_state(
    egress_bps: i64,
    ingress_bps: i64,
    sample: &super::models::BandwidthSample,
    is_recent: bool,
) -> NodeActivityState {
    if !is_recent && egress_bps == 0 && ingress_bps == 0 {
        return NodeActivityState::Idle;
    }
    if egress_bps == 0 && ingress_bps == 0 {
        return NodeActivityState::Idle;
    }
    if sample.loss_rate > 0.05 {
        return NodeActivityState::Degraded;
    }
    let util = if sample.btl_bw_bps > 0 {
        egress_bps as f64 / sample.btl_bw_bps as f64
    } else {
        0.0
    };
    if util >= 0.5 {
        NodeActivityState::Busy
    } else {
        NodeActivityState::Active
    }
}

fn compute_node_state(transports: &[TransportActivitySnapshot]) -> NodeActivityState {
    if transports.is_empty() {
        return NodeActivityState::Offline;
    }
    if transports.iter().any(|t| t.state == NodeActivityState::Degraded) {
        return NodeActivityState::Degraded;
    }
    if transports.iter().any(|t| t.state == NodeActivityState::Busy) {
        return NodeActivityState::Busy;
    }
    if transports.iter().any(|t| t.state == NodeActivityState::Active) {
        return NodeActivityState::Active;
    }
    if transports.iter().all(|t| t.state == NodeActivityState::Offline) {
        return NodeActivityState::Offline;
    }
    NodeActivityState::Idle
}

fn offline_snapshot() -> NodeActivitySnapshot {
    NodeActivitySnapshot {
        state: NodeActivityState::Offline,
        ingress_bps: 0,
        egress_bps: 0,
        active_peers: 0,
        active_transports: 0,
        transports: Vec::new(),
        primary_transport_name: None,
        timestamp: SystemTime::now(),
    }
}

fn now_ms() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or(Duration::ZERO)
        .as_secs_f64()
        * 1000.0
}
